File: finalAssignmentBackend/src/main/java/com/tutict/finalassignmentbackend/config/login/jwt/JwtAuthenticationFilter.java
package com.tutict.finalassignmentbackend.config.login.jwt;

import com.tutict.finalassignmentbackend.config.login.jwt.TokenProvider;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private final TokenProvider tokenProvider;

    public JwtAuthenticationFilter(TokenProvider tokenProvider) {
        this.tokenProvider = tokenProvider;
    }

    @Override
    protected void doFilterInternal(@NotNull HttpServletRequest request, @NotNull HttpServletResponse response, @NotNull FilterChain filterChain)
            throws ServletException, IOException {
        String jwt = getJwtFromRequest(request);
        log.info("Extracted JWT from request: {}", jwt);

        if (jwt != null && tokenProvider.validateToken(jwt)) {
            String username = tokenProvider.getUsernameFromToken(jwt);
            List<String> roles = tokenProvider.extractRoles(jwt);
            log.info("JWT validated. Username: {}, Roles: {}", username, roles);

            List<SimpleGrantedAuthority> authorities = roles.stream()
                    .map(SimpleGrantedAuthority::new)
                    .collect(Collectors.toList());

            UsernamePasswordAuthenticationToken authentication =
                    new UsernamePasswordAuthenticationToken(username, null, authorities);
            SecurityContextHolder.getContext().setAuthentication(authentication);
            log.info("Authentication set for user: {}", username);
        } else {
            log.warn("Invalid or missing JWT in request: {}", request.getRequestURI());
        }

        filterChain.doFilter(request, response);
    }

    private String getJwtFromRequest(HttpServletRequest request) {
        String bearerToken = request.getHeader("Authorization");
        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
            return bearerToken.substring(7);
        }
        return null;
    }
}

File: final_assignment_backend_go/project/configs/jwt/jwt_authentication_filter.go
package jwt

import (
	"log"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"

	"final_assignment_backend_go/project/configs/auth"
)

// AuthenticationMiddleware 验证 JWT 并将用户信息注入 Gin 上下文
func AuthenticationMiddleware(tokenProvider *auth.TokenProvider) gin.HandlerFunc {
	return func(c *gin.Context) {
		jwt := getJwtFromRequest(c)
		log.Printf("[INFO] Extracted JWT from request: %v", jwt)

		if jwt != "" && tokenProvider.ValidateToken(jwt) {
			username, err := tokenProvider.GetUsernameFromToken(jwt)
			if err != nil {
				log.Printf("[WARN] Failed to extract username: %v", err)
				c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "invalid token"})
				return
			}

			roles, err := tokenProvider.ExtractRoles(jwt)
			if err != nil {
				log.Printf("[WARN] Failed to extract roles: %v", err)
				c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "invalid token"})
				return
			}

			log.Printf("[INFO] JWT validated. Username: %s, Roles: %v", username, roles)

			// 注入用户信息
			c.Set("username", username)
			c.Set("roles", roles)

			log.Printf("[INFO] Authentication set for user: %s", username)
			c.Next()
		} else {
			log.Printf("[WARN] Invalid or missing JWT in request: %s", c.Request.RequestURI)
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		}
	}
}

func getJwtFromRequest(c *gin.Context) string {
	bearer := c.GetHeader("Authorization")
	if bearer != "" && strings.HasPrefix(bearer, "Bearer ") {
		return strings.TrimPrefix(bearer, "Bearer ")
	}
	return ""
}
